### PR TITLE
[pallas] Improve error messages for input_output_aliasing

### PR DESCRIPTION
**Pull request by gnecula**
Currently, we get MLIR verification errors when the inputs and outputs declared to be aliased do not have matching shapes and dtypes. We add a nicer error message that localizes the inputs and outputs in the corresponding PyTrees.

Interestingly, if the output index is out of bounds, there is no MLIR verification error, if there is a single output. This is happening because the output index in `input_output_aliases` is ignored during the lowering to StableHLO when
there is a single output, and thus the output_index is implicitly forced to 0.

Currently, in interpreter mode we get a mix of internal assertion errors when there are errors in input_output_aliasing.

### CODE REVIEW

File: jax/_src/pallas/pallas_call.py [hunk 1]
```diff
@@ -1086,6 +1086,27 @@ def wrapped(*args):
       raise ValueError(
           "The kernel function in a pallas_call should return None. "
           f"Found a PyTree: {f_out_tree}")
+    for i_idx, o_idx in input_output_aliases.items():
+      if not (0 <= i_idx < len(flat_in_avals)):
```
Comment (superbobry) [inline jax/_src/pallas/pallas_call.py:1090]: FYI a neat way to write this is 

```
if i_idx not in range(len(flat_in_avals)):
  ...
```

which matches the wording in the error message :)

File: tests/pallas/pallas_test.py [hunk 2]
```diff
@@ -567,6 +567,56 @@ def test_pallas_call_block_shape_ndim_mismatch(self):
...
+    x = np.arange(8 * 128, dtype=np.int32).reshape((8, 128))
+
+    # Invalid input index
+    with self.assertRaisesRegex(
+        ValueError,
+        "input_output_aliases contains the mapping '2:0' with input index 2 "
+        "outside the range .*"):
+      self.pallas_call(lambda x_ref, y_ref, o1_ref: None,
+                       out_shape=[x],
+                       in_specs=[pl.BlockSpec((1, 1), lambda: (0, 0)),
```
Comment (superbobry) [inline tests/pallas/pallas_test.py:580]: I think you can leave out the specs here and below.

File: tests/pallas/pallas_test.py [hunk 3]
```diff
@@ -567,6 +567,56 @@ def test_pallas_call_block_shape_ndim_mismatch(self):
         "array shape"):
       f(a)
 
+  def test_pallas_call_input_output_aliases_errors(self):
+    x = np.arange(8 * 128, dtype=np.int32).reshape((8, 128))
+
+    # Invalid input index
```
Comment (superbobry) [inline tests/pallas/pallas_test.py:573]: Up to you, but I'd drop the comments, because the regex matching the error message is self-explanatory.

**Review by superbobry — APPROVED** [review]
Nice!

File: jax/_src/pallas/pallas_call.py (hunk 1 above)
Comment (gnecula) [reply jax/_src/pallas/pallas_call.py:1090]: True, but it is too inefficient to materialize the range and check one by one.

Comment (superbobry) [reply jax/_src/pallas/pallas_call.py:1091]: `range` doesn't materialize a `list` in 3.X, it just creates a `range` object, but you're right -- there is still an extra allocation.

Comment (gnecula) [reply jax/_src/pallas/pallas_call.py:1091]: I did not know that `range` objects support O(1) element lookup. I am applying the suggestion.

**Pull request merged**